File: routes/reviews.py
from datetime import datetime, timezone

from flask import Blueprint, request

from database.extensions import db, limiter
from models.review import Review
from utils.auth_helpers import admin_required
from utils.response import ok, created, bad_request, invalid_format, not_found

reviews_bp = Blueprint("reviews", __name__)


@reviews_bp.post("/reviews")
@limiter.limit("10/minute")
def submit_review():
    """
    Submit Ulasan (Publik)
    ---
    tags: [Reviews]
    parameters:
      - in: body
        name: body
        schema:
          type: object
          required: [name, rating, review]
          properties:
            name: {type: string}
            role: {type: string}
            rating: {type: integer}
            review: {type: string}
    responses:
      201:
        description: Ulasan diterima, status Menunggu
    """
    body = request.get_json(silent=True) or {}
    name = (body.get("name") or "").strip()
    rating = body.get("rating")
    review_text = (body.get("review") or "").strip()

    if not name or rating is None or not review_text:
        return bad_request(detail='Field "name", "rating", dan "review" wajib diisi.')
    if not isinstance(rating, int) or not (1 <= rating <= 5):
        return invalid_format(detail="Rating harus berupa integer 1-5.")

    entry = Review(name=name, role=body.get("role"), rating=rating, review=review_text)
    db.session.add(entry)
    db.session.commit()

    return created({"reviewId": entry.id, "status": "Menunggu"},
                   "Ulasan berhasil dikirim. Akan ditampilkan setelah disetujui admin.")


@reviews_bp.get("/reviews")
def get_approved_reviews():
    """
    Get Approved Reviews (Publik)
    ---
    tags: [Reviews]
    parameters:
      - in: query
        name: page
        type: integer
      - in: query
        name: limit
        type: integer
    responses:
      200:
        description: Daftar ulasan yang telah disetujui
    """
    page = request.args.get("page", 1, type=int)
    limit = request.args.get("limit", 10, type=int)

    query = Review.query.filter_by(status="Disetujui").order_by(Review.approved_at.desc())
    pagination = query.paginate(page=page, per_page=limit, error_out=False)
    if not pagination.items:
        return not_found(detail="Belum ada ulasan yang disetujui.")

    all_approved = Review.query.filter_by(status="Disetujui").all()
    avg_rating = round(sum(r.rating for r in all_approved) / len(all_approved), 1) if all_approved else 0

    from flask import jsonify
    from utils.response import envelope
    body = envelope("00", "Data ulasan berhasil diambil.", data=[r.to_dict() for r in pagination.items])
    body["summary"] = {"totalReviews": len(all_approved), "averageRating": avg_rating}
    body["pagination"] = {
        "page": page, "limit": limit, "total": pagination.total,
        "totalPages": pagination.pages,
        "nextPage": page + 1 if pagination.has_next else None,
        "prevPage": page - 1 if pagination.has_prev else None,
    }
    return jsonify(body), 200


@reviews_bp.get("/admin/reviews")
@admin_required
def get_pending_reviews():
    """
    Get Pending Reviews (Admin)
    ---
    tags: [Reviews]
    security: [{Bearer: []}]
    parameters:
      - in: query
        name: status
        type: string
        description: >
          Opsional. Default hanya mengembalikan ulasan "Menunggu" sesuai
          API Contract v3.6. Kirim status=all untuk mengambil SEMUA ulasan
          (Menunggu + Disetujui) -- dipakai tab "Semua" di Admin Dashboard.
          Kirim status=Disetujui untuk hanya ulasan yang sudah disetujui.
    responses:
      200:
        description: Daftar ulasan (default hanya yang menunggu moderasi)
    """
    status_filter = request.args.get("status")

    query = Review.query
    if status_filter and status_filter.lower() != "all":
        query = query.filter_by(status=status_filter)
    elif not status_filter:
        query = query.filter_by(status="Menunggu")
    # status=all -> tidak difilter sama sekali

    rows = query.order_by(Review.submitted_at.desc()).all()
    if not rows:
        return not_found(detail="Tidak ada ulasan yang sesuai.")
    return ok([r.to_dict() for r in rows], "Daftar ulasan berhasil diambil.")


@reviews_bp.put("/admin/reviews/<int:review_id>/approve")
@admin_required
def approve_review(review_id):
    """
    Setujui Ulasan (Admin)
    ---
    tags: [Reviews]
    security: [{Bearer: []}]
    responses:
      200:
        description: Ulasan disetujui dan tampil publik
    """
    entry = Review.query.get(review_id)
    if not entry:
        return not_found(detail="Ulasan tidak ditemukan.")
    entry.status = "Disetujui"
    entry.approved_at = datetime.now(timezone.utc)
    db.session.commit()
    return ok(entry.to_dict(), "Ulasan berhasil disetujui.")


@reviews_bp.delete("/admin/reviews/<int:review_id>")
@admin_required
def delete_review(review_id):
    """
    Hapus Ulasan (Admin)
    ---
    tags: [Reviews]
    security: [{Bearer: []}]
    responses:
      200:
        description: Ulasan dihapus
    """
    entry = Review.query.get(review_id)
    if not entry:
        return not_found(detail="Ulasan tidak ditemukan.")
    db.session.delete(entry)
    db.session.commit()
    return ok(None, "Ulasan dihapus")

File: src/app/store/reviews.ts
// src/app/store/reviews.ts
//
// SEBELUM: localStorage, `add()` langsung approved:true.
// SESUDAH: memanggil backend. PERUBAHAN PERILAKU PENTING:
// ulasan yang baru dikirim berstatus "Menunggu" dan BARU tampil di daftar
// publik setelah admin menyetujuinya lewat tab "Ulasan" di Admin Dashboard
// (endpoint PUT /admin/reviews/{id}/approve) -- ini levih sesuai API
// Contract v3.6 dan sudah didukung oleh AdminDashboard.tsx yang sudah ada
// (tab Ulasan sudah punya tombol setujui/hapus, sebelumnya tidak pernah
// terpakai karena versi localStorage selalu auto-approve).

import { apiFetch, apiFetchFull } from "../lib/api";

export interface Review {
  id: string;
  name: string;
  role: string;
  message: string;
  rating: number;
  createdAt: string;
  approved: boolean;
}

// Bentuk baris dari backend (models/review.py -> to_dict())
interface BackendReviewRow {
  id: number;
  name: string;
  role: string | null;
  rating: number;
  review: string;
  status?: string;
  submittedAt?: string | null;
  approvedAt: string | null;
}

function fromBackend(row: BackendReviewRow): Review {
  return {
    id: String(row.id),
    name: row.name,
    role: row.role ?? "Pengguna",
    message: row.review,
    rating: row.rating,
    createdAt: (row.approvedAt ?? row.submittedAt ?? "").slice(0, 10),
    approved: row.status ? row.status === "Disetujui" : true,
  };
}

export const reviewsDb = {
  /** Admin: SEMUA ulasan (Menunggu + Disetujui) -- dipakai tab "Semua" di dashboard. */
  async getAll(): Promise<Review[]> {
    try {
      const rows = await apiFetch<BackendReviewRow[]>("/admin/reviews?status=all");
      return rows.map(fromBackend);
    } catch (err: any) {
      if (err?.code === "11") return [];
      throw err;
    }
  },

  /** Publik: hanya ulasan yang sudah disetujui (ditampilkan di halaman utama). */
  async getApproved(): Promise<Review[]> {
    try {
      const envelope = await apiFetchFull<BackendReviewRow[]>("/reviews?limit=50", { auth: false });
      return (envelope.data ?? []).map(fromBackend);
    } catch (err: any) {
      if (err?.code === "11") return [];
      throw err;
    }
  },

  /** Publik: submit ulasan baru. Status awal "Menunggu" -- BUKAN langsung tampil. */
  async add(data: Omit<Review, "id" | "createdAt" | "approved">): Promise<Review> {
    const result = await apiFetch<{ reviewId: number; status: string }>("/reviews", {
      method: "POST",
      auth: false,
      body: { name: data.name, role: data.role, rating: data.rating, review: data.message },
    });
    return {
      id: String(result.reviewId),
      name: data.name,
      role: data.role,
      message: data.message,
      rating: data.rating,
      createdAt: new Date().toISOString().slice(0, 10),
      approved: false, // sesuai backend: status awal selalu "Menunggu"
    };
  },

  async approve(id: string): Promise<void> {
    await apiFetch(`/admin/reviews/${id}/approve`, { method: "PUT" });
  },

  async delete(id: string): Promise<void> {
    await apiFetch(`/admin/reviews/${id}`, { method: "DELETE" });
  },
};
